teps/0041-move-image-entrypoint-lookup-to-the-taskrun-pod.md

```diff
@@ -0,0 +1,161 @@
+By doing this, we can use the service account that user provided for the TaskRun
+to query the entrypoint of the image that is running on this pod. This will also
+avoid the authentication problem when requests come from different pods on some
+cloud providers and the confusing retricted access to the image when command is not
+specified.
+
+### Goals
+
+- Move the image entry point lookup logic altogether to the entry point binary on the TaskRun pod
+
+### Non-Goals
+
+- Implement a kubernetes api to query the metadata of a image using kubelet
+
+### Use Cases
+
+TaskRun users encounter permission denied error if they don't specify command in
+their container specs on some cloud providers. The error happens when TaskRun
+controller pod tries to pull the private images. Once they specify command for
+the private images, the private images can be pulled successfully. This bug can
+be solved by implementing this TEP.
+
+## Proposal
+
+This TEP proposes to move the image entry point lookup logic altogether to the
+entry point binary on the TaskRun pod. The related logic is going to be
+implemented in the particular TaskRun pod instead of Tekton controller pod. This
+makes things more reasonable and clearer. And the time window between fetching
+image and pod scheduling is also smaller than before.
+
+Even though we are already in the image we want to query the entrypoint from
+after the architecture changing, the image metadata still needs to be fetched
+from `containerregistry`. The ideal solution should be fetching image config
+from the `kubelet` or underlying runtime straightly. However, the `kubelet` API
+that is capable to do this does not exist for now.
+
+### Add Flags to EntryPoint Binary
+
+The entrypoint binary should resolve the image entrypoint or command when the
+step container didn't specify a command. In order to find the image, three
+paramters could be passed to the binary - `image`, `taskrun_namespace`, and
+`taskrun_service_account`. `image` is used to reference the image.
+`taskrun_namespace` and `taskrun_service_account` are used to retrieve the
+authentication to pull the image.
+
+```go
+image = flag.String("image", "", "If specified, pull image and query image entrypoint")
+taskrunNamespace = flag.String("taskrun_namespace", "", "If image specified, use this namespace to pull image")
+taskrunServiceAccount = flag.String("taskrun_service_account", "", "If image specified, use this service account to pull image")
+```
+
+The parameter `entrypoint` must be empty when `image` is specified. The command
+to be run will look like this when step command isn't specified:
+
+```script
+/tekton/tools/entrypoint -wait_file '/tekton/downward/ready' -wait_file_content\
+-post_file '/tekton/tools/0' -termination_path '/tekton/termination'\
+-image '<image_path>' -taskrun_namespace 'default' -taskrun_service_account 'default'
+```
+
+### Resolve Image in EntryPoint Binary
```

Comment:
Part of what the image metadata lookup does is resolve images to their SHA digests, which seems useful for auditing. We should figure out how we can retain this information / bubble it back up to the controller.